src/condor_utils/wire_ad_buffer.cpp

```diff
@@ -0,0 +1,520 @@
+			// fast tricks only work with envelope expr trees
+			fast_tricks = false;
+		}
+	}
+
+	// if we have not been asked to use fast parsing tricks
+	// we can just call the classad update from pickle functions
+	if ( ! fast_tricks) {
+		if (whitelist) {
+			return ad.Update(stmHash, *whitelist, false);
+		}
+		return ad.UpdateViaCache(stmHash);
+	}
+
+	// can't use ad.update because of fast tricks
+	// so iterate the hash here
+
+	int numExprs = WireClassadBuffer::read_hash(stmHash);
+	if (numExprs <= 0) {
+		return false;
+	}
+
+	classad::ClassAdParser parser;
+	parser.SetOldClassAd(true);
+
+	std::string attr; // TODO: remove this when we can insert into a classad froma string_view
+
+	std::string_view lhs;
+	binary_span expr;
+	classad::ExprTree::NodeKind kind;
+
+	for (int ix = 0; ix < numExprs; ++ix) {
+		if ( ! WireClassadBuffer::next_pair(stmHash, lhs, expr, kind) || lhs.empty() || expr.empty()) {
+			dprintf(D_ALWAYS, "updateAdFromRaw next_pair did not return anything to insert\n");
+			return false;
+		}
+
+		// TODO: remove this when we can insert into classads using a string_view
+		attr.assign(lhs);
+
+		// honor whitelist if any
+		if (whitelist && ! whitelist->count(attr)) {
+			continue;
+		}
+
+		classad::ExprStream stmExpr(expr);
+		if (kind != classad::ExprTree::EXPR_ENVELOPE) {
+			classad::ExprTree * tree = nullptr;
+			if (stmExpr.readNullableExpr(tree)) {
+				ad.Insert(attr, tree);
+			} else {
+				dprintf(D_ALWAYS, "updateAdFromRaw next_pair rhs is not a pickled expr\n");
+			}
+			continue;
+		}
+
+		// rhs is an Envelope, which is basically an unparsed expression string
+		unsigned char ct2;
+		std::string_view rhs;
+		stmExpr.readByte(ct2);
+		stmExpr.readSizeString(rhs, ct2 == classad::ExprStream::EsEnvelope);
+
+		// Fast tricks pre-parses the right hand side when it is detected as a simple literal
+		// this is faster than letting the classad parser parse it (as of 8.7.0) and also
+		// uses less memory than letting the classad cache see it since literal nodes are the same
+		// size as envelope nodes.
+		//
+		bool inserted = false;
+		if (fast_tricks) {
+			if (rhs.back() == '\0') rhs.remove_suffix(1);  // size should not include the trailing \0
+			classad::Literal* lit = fastParseSomeClassadLiterals(rhs.data(), rhs.size(), always_cache_string_size);
+			if (lit) {
+				inserted = ad.InsertLiteral(attr, lit);
+			}
+		}
+
+		if (inserted) {
+		} else {
+			// we can't cache nested classads or lists, so just parse and insert them
+			bool cache = use_cache && (rhs.front() != '[' && rhs.front() != '{');
+			if (cache) {
+				inserted = ad.InsertViaCache(attr, rhs.data());
+			} else {
+				ExprTree *tree = parser.ParseExpression(rhs.data());
+				if (tree) {
+					inserted = ad.Insert(attr, tree);
+				}
+			}
+		}
+		if ( ! inserted) {
+			dprintf(D_ALWAYS, "updateAdFromRaw FAILED to insert %s\n", rhs.data());
```

Comment:
See previous dprintf() comment in this function.

Reply:
see above